### DISCUSSION TITLE
ValueError: 0x82 is not implemented in snap7.types

### DISCUSSION THREAD

**Discussion by maxitaxi13** [gijzelaerr/python-snap7#299] [Q&A] [answered]
I made a simple sample code trying to read a value from Q0.0 on PLC S1212C.

```
import snap7
from snap7.util import *
from snap7.types import *
import time
import cv2
import numpy


#POVEZAVA S KRMILNIKOM
PLC=snap7.client.Client()
PLC.connect("192.168.0.1",0,1)

if PLC.get_connected():
    print("connected")
else:
    print("NOT connected")

time.sleep(1)
data=PLC.read_area(0x82, 0, 0, 1)
c=get_bool(data,0,0)

```

The code works just fine on a different computer where I have already installed snap7. On my ne computer i get the error

```
ValueError(f"{area} is not implemented in snap7.types")
ValueError: 130 is not implemented in snap7.types
```

I tried installing it using pip install and the instructions on https://python-snap7.readthedocs.io/en/latest/installation.html
but nothing worked, does anybody have any idea on what to do.

**Comment by swamper123** [comment, marked as answer]
Depending on which version you use, the solution is a bit different. My answer is for version 1.1 (last release):

The install should be finde, if I interpret your error right (hint: debugging line for line and see where it breaks).
It sounds that `PLC.read_area() `is the line who makes trouble.

I guess it's your first parameter in `PLC.read_area()`. It should be from type Areas, not the number directly. So in your code it should look like:

```python
data = PLC.read_area(Areas.PE, 0, 0, 1)
```

You may want to take a look in the comments in the method as well:
```

        Example:
            >>> import snap7
            >>> client = snap7.client.Client()
            >>> client.connect("192.168.0.1", 0, 0)
            >>> buffer = client.read_area(snap7.types.Areas.DB, 1, 10, 4)  # Reads the DB number 1 from the byte 10 to the byte 14.
```

**Reply by swamper123** [reply]
And a Python style hint: You shouldn't do `from XYZ import *`. 😉 
It doesn't look nice and you don't know from where what stuff is imported (and you avoid some traps as well).

**Reply by maxitaxi13** [reply]
Thank you so much, It works now. I figuered it was something like that. I even went to the types. py and found this.  
`
S7AreaPE = 0x81
S7AreaPA = 0x82
S7AreaMK = 0x83
S7AreaDB = 0x84
S7AreaCT = 0x1C
S7AreaTM = 0x1D`

But I dont understand why Areas.PA, where did you find that info.

Thank you so much again.

**Reply by swamper123** [reply]
> But I dont understand why Areas.PA, where did you find that info.

Because I am contributing to that repo for a while. 😉 

If you go to the line in the repo:
https://github.com/gijzelaerr/python-snap7/blob/7aea4f62838c03135accd87c5434c662079ad193/snap7/client.py#L368
You will see, that the parameter has to be from Type "Areas" and "Areas" is located in snap7.types.

Since you made:
 ```python
from snap7.types import *
```
You imported everything from snap7.types, the class `Area `as well. So you need to pass the right name of that.

The stuff you found in types  `S7AreaPE = 0x81 ` etc. is partly a leftover from older versions, where this has been used.... 

@lautarodapin are they still needed somewhere? I mean, that's what the Enum is for, isn't it?

**Reply by maxitaxi13** [reply]
Awsome, thank you :)
